src/kde/kde.t.hh

```diff
@@ -0,0 +1,145 @@
+ * \author Mathew Cleveland
+ * \date   November 10th 2020
+ * \brief  Templated KDE functions for various dimensions and coordinate
+ * systems
+ * \note   Copyright (C) 2018-2020 Triad National Security, LLC.
+ *         All rights reserved. */
+//------------------------------------------------------------------------------------------------//
+
+#ifndef kde_kde_t_hh
+#define kde_kde_t_hh
+
+#include "kde.hh"
+#include <iostream>
+#include <math.h>
+#include <numeric>
+
+namespace kde {
+
+/*!
+ * reconstruction 
+ * \brief
+ *
+ * Cartesian geometry reconstruction of a 1D distribution
+ *
+ * \param[in] distribution 
+ * \param[in] position
+ * \param[in] band_width
+ * \param[inout] result returned final local function distribution
+ *
+ */
+template <>
+template <>
+std::vector<double> kde<kde_coordinates::CART>::reconstruction<1>(
+    const std::vector<double> &distribution, const std::vector<std::array<double, 3>> &position,
+    const std::vector<std::array<double, 3>> &band_width, const bool domain_decomposed) const {
+  const int64_t local_size = distribution.size();
+  Check(static_cast<int64_t>(position.size()) == local_size);
+  Check(static_cast<int64_t>(band_width.size()) == local_size);
+
+  int64_t size = local_size;
+  int64_t global_lower_bound = 0;
+  int64_t global_upper_bound = local_size;
+  double global_conservation = std::accumulate(distribution.begin(), distribution.end(), 0.0);
+  // minimize global values and only allocate them in DD problems
+  std::vector<double> global_distribution;
+  std::vector<double> global_x_position;
+  if (domain_decomposed) {
+    // calculate global off sets
+    int n_ranks = rtt_c4::nodes();
+    std::vector<int64_t> rank_size(rtt_c4::nodes(), 0);
+    rank_size[rtt_c4::node()] = local_size;
+    rtt_c4::global_sum(rank_size.data(), n_ranks);
+    size = std::accumulate(rank_size.begin(), rank_size.end(), 0.0);
+    std::vector<int64_t> accum_rank_size(rank_size);
+    std::partial_sum(rank_size.begin(), rank_size.end(), accum_rank_size.begin());
+
+    if (rtt_c4::node() > 0) {
+      global_lower_bound = accum_rank_size[rtt_c4::node() - 1];
+      global_upper_bound = accum_rank_size[rtt_c4::node()];
+    }
+
+    // set up global arrays
+    global_distribution.resize(size, 0.0);
+    global_x_position.resize(size, 0.0);
+
+    // build up global positions
+    for (int i = 0; i < local_size; i++) {
+      global_x_position[i + global_lower_bound] = position[i][0];
+      global_distribution[i + global_lower_bound] = distribution[i];
+    }
+
+    rtt_c4::global_sum(global_x_position.data(), size);
+    rtt_c4::global_sum(global_distribution.data(), size);
+    rtt_c4::global_sum(global_conservation);
+  }
+
+  std::vector<double> result(local_size, 0.0);
+  std::vector<double> normal(local_size, 0.0);
+
+  // now apply the kernel to the local ranks
+  for (int i = 0; i < local_size; i++) {
+    const double x0 = position[i][0];
+    const double h = band_width[i][0];
```

Comment:
I think it makes sense to store 1/h instead of h since it's always the divisor.

Reply:
This pushes off the check for divide by zero that I would have had also. I like this idea. I'll change it.